core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,274 @@
+
+    while (remainingBytesToRead > 0) {
+      byte[] plainBlock = decryptNextBlock(isLastBlockInStream);
+
+      int remainingBytesInBlock = plainBlock.length - currentOffsetInPlainBlock;
+      boolean finishTheBlock = remainingBytesToRead >= remainingBytesInBlock;
+      int bytesToCopy = finishTheBlock ? remainingBytesInBlock : remainingBytesToRead;
+      System.arraycopy(plainBlock, currentOffsetInPlainBlock, b, resultBufferOffset, bytesToCopy);
+      remainingBytesToRead -= bytesToCopy;
+      resultBufferOffset += bytesToCopy;
+      currentOffsetInPlainBlock += bytesToCopy;
+
+      boolean endOfStream = isLastBlockInStream && finishTheBlock;
+
+      if (endOfStream) {
+        break;
+      }
+
+      if (finishTheBlock) {
+        currentBlockIndex++;
+        currentOffsetInPlainBlock = 0;
+        isLastBlockInStream = (currentBlockIndex + 1 == numberOfBlocks);
+      }
+    }
+
+    plainStreamPosition += len - remainingBytesToRead;
+    return len - remainingBytesToRead;
+  }
+
+  @Override
+  public void seek(long newPos) throws IOException {
+    if (newPos < 0) {
+      throw new IOException("Negative new position " + newPos);
+    } else if (newPos > plainStreamSize) {
+      throw new EOFException(
+          "New position " + newPos + " exceeds the max stream size " + plainStreamSize);
+    }
+
+    currentBlockIndex = Math.toIntExact(newPos / plainBlockSize);
+    currentOffsetInPlainBlock = Math.toIntExact(newPos % plainBlockSize);
+    plainStreamPosition = newPos;
+  }
+
+  @Override
+  public long skip(long n) {
+    if (n <= 0) {
+      return 0;
+    }
+
+    if (plainStreamPosition == plainStreamSize) {
+      return 0;
+    }
+
+    long newPosition = plainStreamPosition + n;
+
+    if (newPosition > plainStreamSize) {
+      long skipped = plainStreamSize - plainStreamPosition;
+      try {
+        seek(plainStreamSize);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+      return skipped;
+    }
+
+    try {
+      seek(newPosition);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    return n;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return plainStreamPosition;
+  }
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    sourceStream.close();
+    currentDecryptedBlock = null;
+  }
+
+  static long calculatePlaintextLength(long sourceLength, int plainBlockSize) {
+    long netSourceFileLength = netSourceFileLength(sourceLength);
+
+    if (netSourceFileLength == 0) {
+      return 0;
+    }
+
+    int cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+    int numberOfFullBlocks = Math.toIntExact(netSourceFileLength / cipherBlockSize);
+    int cipherBytesInLastBlock =
+        Math.toIntExact(netSourceFileLength - numberOfFullBlocks * cipherBlockSize);
```

Comment:
Since the stream length is a long, I don't think it makes sense to make `numberOfFullBlocks` in int. The multiplication can clearly produce a long anyway so this should remove the `toIntExact` above. Then everything is a long until `cipherBytesInLastBlock`.